core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -363,6 +344,10 @@ public RetryOnReplicaException(StreamInput in) throws IOException {
     }
 
     private final class AsyncReplicaAction extends AbstractRunnable {
+        /**
+         * The number of operations remaining before we can reply. See javadoc for {@link #operationComplete()} more.
+         */
+        private final AtomicInteger operationsUntilReply = new AtomicInteger(2);
```

Comment:
I think this should be explicit. we should have something like:
```
private volatile boolean onPrimaryDone = false;
private volatile boolean onReplicaDone = false;

private synchronized void doneOnPrimary() {
  assert onPrimaryDone == false;
  onPrimaryDone = true;
}
```

that way we know if we set it twice or if what is going wrong if we miss one etc?